flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/endpoint/hive/util/ThriftObjectConversionsTest.java

```diff
@@ -34,4 +107,249 @@ public void testConvertSessionHandle() {
+                        .expectSqlType(Types.FLOAT)
+                        .expectValue(1024.0),
+                DataTypeSpec.newSpec()
+                        .forType(DOUBLE())
+                        .forValue(2048.1024)
+                        .expectSqlType(Types.DOUBLE),
+                DataTypeSpec.newSpec()
+                        .forType(DECIMAL(9, 6))
+                        .forValue(DecimalData.fromBigDecimal(new BigDecimal("123.456789"), 9, 6))
+                        .expectSqlType(Types.DECIMAL)
+                        .expectValue("123.456789"),
+                DataTypeSpec.newSpec()
+                        .forType(STRING())
+                        .forValue(StringData.fromString("Hello World"))
+                        .expectSqlType(Types.VARCHAR)
+                        .expectValue("Hello World"),
+                DataTypeSpec.newSpec()
+                        .forType(BYTES())
+                        .forValue("Flink SQL Gateway".getBytes(StandardCharsets.UTF_8))
+                        .expectSqlType(Types.BINARY)
+                        .expectValue(
+                                new String("Flink SQL Gateway".getBytes(StandardCharsets.UTF_8)),
+                                "Flink SQL Gateway".getBytes(StandardCharsets.UTF_8)),
+                DataTypeSpec.newSpec()
+                        .forType(DATE())
+                        .forValue((int) LocalDate.parse("2022-02-22").toEpochDay())
+                        .expectSqlType(Types.DATE)
+                        .expectValue("2022-02-22"),
+                DataTypeSpec.newSpec()
+                        .forType(TIMESTAMP(4))
+                        .forValue(
+                                TimestampData.fromLocalDateTime(
+                                        LocalDateTime.parse("2020-05-11T12:00:12.1234")))
+                        .expectSqlType(Types.TIMESTAMP)
+                        .expectValue("2020-05-11 12:00:12.1234"),
+                DataTypeSpec.newSpec()
+                        .forType(DataTypes.MAP(DataTypes.STRING(), DataTypes.STRING()))
+                        .forValue(mapData)
+                        .expectSqlType(Types.JAVA_OBJECT)
+                        .expectValue("{\"Hello\":\"World\",\"World\":\"Hello\"}")
+                        .expectNullValue("null"),
+                DataTypeSpec.newSpec()
+                        .forType(
+                                DataTypes.ARRAY(
+                                        DataTypes.MAP(DataTypes.STRING(), DataTypes.STRING())))
+                        .forValue(new GenericArrayData(new Object[] {mapData, mapData}))
+                        // Hive uses STRING type
+                        .expectSqlType(Types.VARCHAR)
+                        .expectValue(
+                                "[{\"Hello\":\"World\",\"World\":\"Hello\"},{\"Hello\":\"World\",\"World\":\"Hello\"}]")
+                        .expectNullValue("null"));
```

Comment:
Why there is a null value for the complex types?

Reply:
Every DataType test contains two parts:

1. Serialization can get the expected value for the specific value.
2. Serialization can serialize the null value.

Hive serializes the null value for the complex type to the string "null"  rather than null[1].

[1] https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/SerDeUtils.java#L317